ASM feature (#96)
autonity/docs.autonity.org: content/docs/concepts/asm/_index.md

---
title: "Auton Stabilization Mechanism (ASM)"
linkTitle: "Auton Stabilization Mechanism (ASM)"
weight: 10
description: >
  Auton Stabilization Mechanism: elements, the functions they perform, and the lifecycle for Auton and Newton supply.
---

## Overview

This section describes the [Auton Stabilization Mechanism (ASM)](/glossary/#asm) and protocol. It details the elements comprising the mechanism, the functions the mechanism provides to compute and maintain a stable price for [Auton (ATN)](/glossary/#auton), and the lifecycle for Auton and [Newton (NTN)](/glossary/#newton) supply.

> For Auton stabilization control Autonity implements a [CDP](/glossary/#cdp)-based stabilization mechanism. Users take out CDP's, depositing Collateral Token (NTN) to borrow Auton at interest. As CDP's are taken out and repaid Collateral Token (NTN) and ATN are removed and returned to circulation, bringing equilibrium to supply and demand.

Reviewer: ```suggestion
For Auton stabilization control Autonity implements a [CDP](/glossary/#cdp)-based stabilization mechanism. Users take out CDPs, depositing Collateral Token (NTN) to borrow Auton at interest. As CDPs are taken out and repaid Collateral Token (NTN) and ATN are removed and returned to circulation, bringing equilibrium to supply and demand.
```

> CDP are maintained according to collateralization and liquidation ratios that set debt to collateral ratio to keep a CDP in good health. Auton is minted and burned as CDP's pass through their lifecycle, i.e. are taken out, repaid, withdrawn, and liquidated.

Reviewer: ```suggestion
CDP are maintained according to collateralization and liquidation ratios that set debt to collateral ratio to keep a CDP in good health. Auton is minted and burned as CDPs pass through their lifecycle, i.e. are taken out, repaid, withdrawn, and liquidated.
```

Elasticity in supply and demand for Auton is absorbed by dynamically adjusting CDP incentives to increase and decrease Auton borrowing costs when Auton price moves above or below its Stabilization Target the [Auton Currency Unit (ACU)](/glossary/#acu).

## ASM identifiers and accounts

The ASM functions with two identities for cryptographic security: the CDP owner and the stabilization protocol contract.

> Stabilization Contract calls to mint and burn Auton as CDP's are interacted with to borrow and repay Auton are  restricted to the Stabilization Contract address, the '`stabilizer`' protocol address.

Reviewer: ```suggestion
Stabilization Contract calls to mint and burn Auton as CDPs are interacted with to borrow and repay Auton are  restricted to the Stabilization Contract address, the '`stabilizer`' protocol address.
```

The ACU currency basket is composed of 7 free-floating currencies:

- AUD - Australian Dollar
- CAD - Canadian Dollar
- EUR - Euro
- GBP - British Pound Sterling
- JPY - Japanese Yen
- SEK - Swedish Krona
- USD - United States Dollar

Each currency's quantity in the basket is computed to provide a currency basket with minimal total variance with respect to its underlying currencies. The index value then has minimal volatility with respect to variance from individual currency fluctuations.

ACU is used as the _stabilization target_ for Auton price, see _[Stabilization](/concepts/asm/#stabilization)_.

The value of ACU can be computed at any time in terms of exchange rates. ACU is computed based on the end-of-day time for the most recent 11 calendar years of data for each currency in the basket.

{{% alert title="Info" %}}
A 365 calendar day is used, end-of-day is 17:00 GMT. Price data is sourced from off-chain by the validator [oracle network](/concepts/oracle-network/) and retrieved from the [Oracle Contract](/concepts/architecture/#autonity-oracle-contract) on-chain.

> {{% /alert %}}

Reviewer: Focusing on these three statements:

1. ACU is computed based on the end-of-day time for the most recent 11 calendar years of data for each currency in the basket.
2. A 365 calendar day is used, end-of-day is 17:00 GMT.
3. Price data is sourced from off-chain by the validator [oracle network](/concepts/oracle-network/) and retrieved from the [Oracle Contract](/concepts/architecture/#autonity-oracle-contract) on-chain.

These two paragraphs conflate:

- How the ACU basket quantities have been computed (1 and 2)
- How the ACU index _value_ is computed (3)

Would suggest making distinction between the two.

Author: Yes, agree.  I've drafted a rewording per beneath but want to check with @rtedwardchen before changing it in the PR.

Rewording:

>The value of ACU can be computed at any time in terms of exchange rates. This is computed in 2 logical steps:
>
>1. ACU basket quantity weights. The quantity of each currency in the basket is computed based on the end-of-day time for the most recent 11 calendar years of rate data for each currency. A 365 calendar day is used, end-of-day is 17:00 GMT.
>2. ACU index value. The ACU value is computed from price data for the basket currencies. Price data is sourced from off-chain by the validator [oracle network](/concepts/oracle-network/) and retrieved from the [Oracle Contract](/concepts/architecture/#autonity-oracle-contract) on-chain.

Author: OK, after a call with @rtedwardchen we've reworded the ACU section to just this:

---
#### ACU

The [Auton Currency Unit (ACU)](/glossary/#acu) is a currency basket from which an index value is computed. This index value is then used as the _stabilization target_ for Auton price, see _[Stabilization](/concepts/asm/#stabilization)_.

The ACU currency basket is composed of 7 free-floating currencies:

- AUD - Australian Dollar
- CAD - Canadian Dollar
- EUR - Euro
- GBP - British Pound Sterling
- JPY - Japanese Yen
- SEK - Swedish Krona
- USD - United States Dollar

Each currency's quantity in the basket is computed to provide a currency basket with minimal total variance with respect to its underlying currencies. The index value then has minimal volatility with respect to variance from individual currency fluctuations.

The index value of ACU can be computed at any time in terms of exchange rates. The index value is computed from price data for each of the basket currencies. Price data is sourced from off-chain by the validator [oracle network](/concepts/oracle-network/) and retrieved from the [Oracle Contract](/concepts/architecture/#autonity-oracle-contract) on-chain.

The value is recomputed at the end of each [oracle voting round](/concepts/oracle-network/#voting-rounds) after new price data for the basket currencies has been computed by the [oracle network](/concepts/oracle-network/).

Public functions can be called to return the ACU value, the currency pair symbols in the basket, and the basket quantities, see [ACU Contract Interface](/reference/api/asm/acu/).

Modifying the currency basket is restricted to the governance account. See the Governance and Protocol Only Reference, [`modifyBasket()`](/reference/api/aut/op-prot/#modifybasket-acu-contract).

---

This is much clearer...

Author: This was done in commit 5f730dfbd65d67c907991af910e6ff1ca3cb7bba.

USD is used as the numeraire over the data. The quantity of each currency in the basket is then computed based on a weighting that aims to minimize its variance with respect to the basket and using an initial target value of 1 USD for the total value of the basket. Basket quantities fixed over time. The value of ACU at any time can be computed in terms of exchange rates.

To illustrate supposing the ACU is computed at 17:00 GMT today:

- Collect end-of-day (17:00 GMT) data for each ACU constituent currency for the preceding 11 calendar years to today.
- Compute minimization, weights, and basket quantities using the data
- Compute the ACU value for today using the data.

> When the computed weights are determined to no longer be optimal, they are updated. The "recompute time" is then the close of the most recent calendar day in the past and again based on the preceding 11 years of calendar data from that time point the new basket quantities are computed.

Reviewer: This will need careful review as well as I'm not sure its entirely accurate, or it risks creating confusion.

Author: It has been reviewed by @rtedwardchen but I will go through it again with him when I discuss the proposed rewording in the preceding comment.

Author: OK, after a call with @rtedwardchen we reworded then decided to comment it out/remove it. (See the revised ACU section text in preceding comment https://github.com/autonity/docs.autonity.org/pull/96#discussion_r1361934747 without it)

---

The corrected wording for the basket before commenting out was:

{{% alert title="A note on ACU basket quantities" color="info" %}}
The quantity of each currency in the basket is computed based on the end-of-day time for the most recent 11 calendar years of rate data for each currency. A 365 calendar day is used, end-of-day is 17:00 GMT.

USD is used as the numeraire over the data. The quantity of each currency in the basket is then computed based on a weighting that aims to minimize its variance with respect to the basket and using an initial target value of 1 USD for the total value of the basket. Basket quantities fixed over time. The value of ACU at any time can be computed in terms of exchange rates.

When the computed weights are determined to no longer be optimal, the basket quantities are recomputed. The "recompute time" is then the close of the most recent calendar day in the past and again based on the preceding 11 years of calendar data from that time point.

The new basket quantities are computed by:

- Collecting end-of-day (17:00 GMT) data for each ACU constituent currency for the preceding 11 calendar years to today.
- Compute minimization, weights, and basket quantities using the data.
{{% /alert %}}

Author: This was done in commit 5f730dfbd65d67c907991af910e6ff1ca3cb7bba.

Modifying the currency basket is restricted to the governance account. See the Governance and Protocol Only Reference, [`modifyBasket()`](/reference/api/aut/op-prot/#modifybasket-acu-contract).

#### Supply control

The Supply Control contract controls the supply of Auton in circulation on the network. The contract is called by the [Stabilization](/concepts/asm/#stabilization) contract to mint and burn Auton as CDP's are opened and Auton borrowed and repaid. See the Governance and Protocol Only Reference, [`mint()`](/reference/api/aut/op-prot/#mint-supply-control-contract) and [`burn()`](/reference/api/aut/op-prot/#burn-supply-control-contract).

Public functions to return the total supply of Auton and the amount of Auton available for minting can be called, see [Supply Control Contract Interface](/reference/api/asm/supplycontrol/).

#### Stabilization

The Stabilization Contract maintains a ledger of CDPs and calls the Supply Control Contract to mint and burn Auton as collateral token is deposited or withdrawn and borrowing repaid.

> The Stabilization Contract maintains an _Internal Balance Sheet_ ledger of CDP debt, recording for each CDP:

Reviewer: I don't think ledger is an accurate description, and I would not advise use of IBS:

```suggestion
The Stabilization Contract maintains a record of CDPs and calls the Supply Control Contract to mint and burn Auton as collateral token is deposited or withdrawn and borrowing repaid. For each CDP, the Stabilization Contract records:
```

- `timestamp`: the timestamp of the last borrow or repayment.
- `collateral`: the collateral deposited with the Stabilization Contract.
- `principal`: the principal debt outstanding as of `timestamp`.
- `interest`: the interest debt that is due at the `timestamp`.

> Stabilization functions by dynamically adjusting CDP incentives.

Reviewer: "functions" is quite an overloaded term when talking about code

```suggestion
The stabilization mechanism operates by dynamically adjusting CDP incentives.
```

Users post Collateral Token to borrow ATN against collateral at the Borrow Rate. The Auton Borrow Rate goes up (down) depending on whether ATN/ACU is below (above) the target exchange rate for ATN/ACU to:

  - Increase ATN borrowing (more supply) when ATN/ACU is above target
  - Decrease ATN borrowing (less supply) when ATN/ACU is below target.

> CDP collateral token is currently restricted to the Autonity staking token NTN. A next generation of the ASM will also accept LNTN as collateral. In this future iteration, stakers who post LNTN collateral will continue to receive their staking awards, eliminating opportunity cost distortions. I.e. lending NTN collateral introduces an opportunity cost of paying interest and losing opportunity to earn staking rewards if bonded.

Reviewer: I'm not sure it is good idea to discuss roadmap/features in the context of this document, so I would omit this:

```suggestion
CDP collateral token is currently restricted to the Autonity staking token NTN but may be extended to other protocol assets.
```

Author: >I'm not sure it is good idea to discuss roadmap/features in the context of this document, 

OK.

- _SCALE_FACTOR_ = `10 ** SCALE`, the multiplier for scaling numbers to the required scale.
- _SECONDS_IN_YEAR_ = `365 days`, a year is assumed to have 365 days for interest rate calculations.   
-->

### Protocol primitives

Essential primitives of ASM are: collateral, exchange rate price data, ACU, and the CDP.

#### Collateral

Auton borrowing is collateralized by depositing collateral token into a CDP. The amount and value of collateral backing a CDP is determined by collateralization and liquidation ratios set in the [ASM configuration](/concepts/asm/#asm-configuration). Failure by a [borrower](/concepts/asm/#stabilization-roles) to maintain these ratios results in a CDP becoming liquidatable. In a liquidation scenario a [liquidator](/concepts/asm/#stabilization-roles) is able to assume the debt position, repay outstanding debt, and receive the position's remaining collateral in return.

Autonity's native protocol asset [Newton (NTN)](/concepts/protocol-assets/newton/) is used as the collateral token.

{{% alert title="Info" %}}
A future iteration of ASM will enable the use of [Liquid Newton (LNTN)](/concepts/protocol-assets/liquid-newton/) as collateral token.

> {{% /alert %}}

Reviewer: I would omit this claim:

```suggestion
```

Oracle price data is computed per the [Oracle protocol](/concepts/oracle-network/#oracle-protocol), updated periodically in [voting rounds](/glossary/#voting-round).

#### ACU

Auton price has the [Auton Currency Unit (ACU)](/glossary/#acu) as the Stabilization Target to which it _mean-reverts_. ACU is an index value computed from a basket of free-floating currencies. Use of a currency basket minimises exposure to an individual currency's FX exchange risk. The index value is computed from the basket, weighted _pro rata_ to each currency's share i.e. _quantity_ in the basket. Basket quantities are set at network genesis and may be modified by governance.

ASM then functions to maintain Auton-to-ACU value, '_mean reverting_' to this value by the CDP stabilization mechanism. 

ACU value is kept current by protocol recomputing the value at the end of each oracle voting round when price data for all the basket currencies is available. See the Protocol Only Reference functions [`modifyBasket()`](/reference/api/aut/op-prot/#modifybasket-acu-contract) [`update()`](/reference/api/aut/op-prot/#update-acu-contract) for more detail.

> #### CDP

Reviewer: This CDP section here is missing crucial details about how these various quantities are computed, ie. formulas for borrow interest rate, borrow limit, how interest is computed, ...

> The CDP functions to manage Auton borrowing and stabilize Auton price by adjusting the CDP borrowing cost. CDP's are operated within strict parameterisation constraints:

Reviewer: ```suggestion
The CDP functions to manage Auton borrowing and stabilize Auton price by adjusting the CDP borrowing cost. CDP's are operated within strict parameterization constraints:
```

---
title: "Auton Stabilization Mechanism (ASM)"
linkTitle: "Auton Stabilization Mechanism (ASM)"
weight: 10
description: >
  Auton Stabilization Mechanism: elements, the functions they perform, and the lifecycle for Auton and Newton supply.
---

## Overview

This section describes the [Auton Stabilization Mechanism (ASM)](/glossary/#asm) and protocol. It details the elements comprising the mechanism, the functions the mechanism provides to compute and maintain a stable price for [Auton (ATN)](/glossary/#auton), and the lifecycle for Auton and [Newton (NTN)](/glossary/#newton) supply.

For Auton stabilization control Autonity implements a [CDP](/glossary/#cdp)-based stabilization mechanism. Users take out CDP's, depositing Collateral Token (NTN) to borrow Auton at interest. As CDP's are taken out and repaid Collateral Token (NTN) and ATN are removed and returned to circulation, bringing equilibrium to supply and demand.

> CDP are maintained according to collateralization and liquidation ratios that set collateral amount and collateral value thresholds to keep a CDP in good health.  Auton is minted and burned as CDP's pass through their lifecycle, i.e. are taken out, repaid, withdrawn, and liquidated.

Reviewer 2: CDP are maintained according to collateralization and liquidation ratios that set **~collateral amount and collateral value thresholds~ debt to collateral ratio** to keep a CDP in good health. Auton is minted and burned as CDP's pass through their lifecycle, i.e. are taken out, repaid, withdrawn, and liquidated.

Reviewer 2: @cmjc I don't see any problem with your description, because the collateral amount and collateral value thresholds are health factors that relate to liquidation ratio. However, IMO debt to collateral ratio may be a more direct description.

Author: Done in commit 6240f03709feec9b14e779a078f0624f50b1aa7d

- *Liquidator (Keeper)*: a user or agent that liquidates a CDP that has brobecome under collateralized, repaying the CDP's outstanding debt and receiving remaining collateral token in return. Liquidator's must have an [account](/glossary/#account) on the Autonity network.
- *Stabilizer (ASM Protocol)*: the `stabilizer` is the protocol account address used for Auton `mint` and `burn` operations by the Stabilization Contract)

### ASM elements

The ASM is composed of 3 system elements implemented as smart contract logic: ACU, Supply Control, Stabilization. ASM contracts are deployed by the protocol at genesis.

#### ACU

The [Auton Currency Unit (ACU)](/glossary/#acu) is a currency basket from which an index value is computed. This index value is then used as the _stabilization target_ for Auton price.

The ACU currency basket is composed of 7 free-floating currencies:

- AUD - Australian Dollar
- CAD - Canadian Dollar
- EUR - Euro
- GBP - British Pound Sterling
- JPY - Japanese Yen
- SEK - Swedish Krona
- USD - United States Dollar

> Each currency's quantity in the basket is computed to provide a currency basket with minimal variance. The index value then has minimal volatility with respect to variance from individual currency fluctuations.

Reviewer 2: Each currency's quantity in the basket is computed to provide a currency basket with minimal **averaged** variance **with respect to its underlying currencies**. The index value then has minimal volatility with respect to variance from individual currency fluctuations.

Reviewer 2: @cmjc For ACU underlying currency quantity computation, it is computed through an optimization process, which aims to minimize the average of all currencies' variance w.r.t. the basket (the variance of the basket itself is not optimized).

In other words, it aims to minimize `AVERAGE(VAR(AUD/ACU), VAR(CAD/ACU), VAR(EUR/ACU), VAR(GBP/ACU), VAR(JPY/ACU), VAR(USD/ACU), VAR(SEK/ACU))`.

Author: Done in commit 6240f03709feec9b14e779a078f0624f50b1aa7d

ACU is used as the _stabilization target_ for Auton price, see _[Stabilization](/concepts/asm/#stabilization)_.

The value of ACU can be computed at any time in terms of exchange rates. ACU is computed based on the end-of-day time for the most recent 11 calendar years of data for each currency in the basket.

{{% alert title="Info" %}}
A 365 calendar day is used, end-of-day is 17:00 GMT. Price data is sourced from off-chain by the validator [oracle network](/concepts/oracle-network/) and retrieved from the [Oracle Contract](/concepts/architecture/#autonity-oracle-contract) on-chain.
{{% /alert %}}

> USD is used as the numeraire over the data. The quantity of each currency in the basket is then computed based on a weighting that minimizes variance and using an initial target value of 1 USD for the total value of the basket. Basket quantities fixed over time. The value of SAC at any time can be computed in terms of exchange rates.

Reviewer 2: USD is used as the numeraire over the data. The quantity of each currency in the basket is then computed based on a weighting that **aims to ~minimizes~ minimize its** variance **with respect to the basket** and using an initial target value of 1 USD for the total value of the basket. Basket quantities fixed over time. The value of SAC at any time can be computed in terms of exchange rates.

Author: I edited SAC to ACU.  

Else, done in commit 6240f03709feec9b14e779a078f0624f50b1aa7d

    - Calls the Stabilization Contract to deposit Collateral Token to a CDP, calling the contract's [`deposit()`(/reference/api/asm/stabilization/#deposit) to deposit collateral.
  - Stabilization Contract creates the CDP, using the ERC20 allowance mechanism to execute the collateral deposit. A CDP object is created, and the [CDP attributes](/concepts/asm/#stabilization) are populated.

- CDP is maintained and serviced. The CDP Owner maintains the CDP, opting to increase or decrease borrowing and deposited collateral within [CDP primitive constraints](/concepts/asm/#cdp). The CDP Owner may:

  - _Borrow_: CDP Owner borrows Auton from CDP against collateral. Constraint checks are applied:
    - the amount borrowed must not exceed the _borrow limit_ for the deposited collateral value
    - the _principal_ debt must meet the _minimum debt requirement_

>     - the borrowing does not make the CDP meet the _liquidation condition_.

Reviewer 2: - the borrowing does not **~make~ decrease** the CDP **~meet the _liquidation condition_~ collateralization ratio below the _minimum collateralization ratio_**.

Reviewer 2: @cmjc  IMO borrowing shall not decrease the CDP's collateralization ratio < minimum collateralization ratio.

Author: Got it. Done in commit 6240f03709feec9b14e779a078f0624f50b1aa7d.
